Delete stale UDP conntrack entries that use hostPort

prameshj opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR introduces a change to delete stale conntrack entries for UDP connections, specifically for udp connections that use hostPort. When the pod listening on that udp port get updated/restarted(and gets a new ip address), these entries need to be flushed so that ongoing udp connections can recover once the pod is back and the new iptables rules have been installed. 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59033

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

prameshj committed:
Remove conntrack entry on udp rule add.

Moved conntrack util outside of proxy pkg
Added warning message if conntrack binary is not found
Addressed review comments.
ran gofmt

prameshj commented:
/assign freehan

prameshj commented:
signed the CNCF CLA

MrHohn commented:
/ok-to-test

prameshj commented:
/retest

prameshj commented:
/test pull-kubernetes-bazel-test

bowei commented on pkg/kubelet/network/hostport/hostport_manager.go:
```diff
@@ -150,6 +159,21 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 	}
+	isIpv6 := utilconntrack.IsIPv6(podPortMapping.IP)
+
+	// Remove conntrack entries just after adding the new iptables rules. Better to do this here than in Remove because
```
I'm find it a little hard to understand this comment. Maybe something like 
```
// Remove conntrack entries just after adding the new iptables rules. If the conntrack entry
// is removed along with the IP tables rule, it can be the case that the packets received by
// the client after IP table rule removal will create a new conntrack entry that will result in
//  blackhole of the traffic.
```

bowei commented on pkg/kubelet/network/hostport/hostport_manager_test.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"k8s.io/api/core/v1"
 	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+	utilexec "k8s.io/utils/exec"
```
don't rename packages unless it is necessary

thockin replied:
This one is pretty common

freehan commented:
/approve

bowei commented:
/assign @dchen1107 

Please reassign as appropriate

freehan commented:
/lgtm

madhusudancs commented on hack/.golint_failures:
```diff
@@ -343,6 +343,7 @@ pkg/ssh
 pkg/util/bandwidth
 pkg/util/config
 pkg/util/configz
+pkg/util/conntrack
```
Please fix the lint warnings by adding comments to the exported names instead of adding this package to the ignore file. I am worried that this package will stay here for ever if we don't fix this now and the fix is not that involved. See the lint failures and suggestions here: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/59286/pull-kubernetes-verify/75380/#verify-golint

madhusudancs commented:
/lgtm

LGTM for the refactor, i.e. `pkg/proxy/util/conntrack*` → `pkg/util/conntrack/...`. Please wait for an LGTM from someone who knows kubelet and networking too.

prameshj commented:
/assign @thockin

freehan commented:
@squeed Probably make sense to adopt this fix to portmap?

dchen1107 commented:
cc/ @adityakali @kubernetes/goog-image for potential issues?

adityakali commented:
IIUC this is invoking `conntrack` tool from host OS? Most distros do not have `conntrack` tool installed by default. Debian & Ubuntu doesn't have it and neither does COS (unless it is invoked from the [toolbox container](https://cloud.google.com/container-optimized-os/docs/how-to/toolbox)).
So unfortunately this will not help.

I will prefer if k8s does not add dependencies on specialized tools in the OS. Better option is to carry dependencies in a separate container similar to how it is done for mount utilities: [Proposal: containerized mount utilities in pods](https://github.com/kubernetes/community/pull/589).

thockin commented on pkg/kubelet/network/hostport/hostport_manager.go:
```diff
@@ -45,6 +48,7 @@ type HostPortManager interface {
 
 type hostportManager struct {
 	hostPortMap map[hostport]closeable
+	conntrack   utilexec.Interface
```
call this `exec` or something - it's not about conntrack

thockin reviewed (commented):
The joy of making better package boundaries is renaming things and fixing APIs!

I propose to:

rename `ClearUDPConntrackForIP` to `ClearToIP` and take "udp" as an argument.

rename `ExecConntrackTool` to `Exec`

rename `ClearUDPConntrackForPort` to `ClearToPort` and take "udp" as an argument

rename `ClearUDPConntrackForPeers` to `ClearToMappedIP` and take "udp" as an argument

These are not great (open to ideas) but a re a minimal API cleanup.  This pkg should probably eventually look more like iptables and ipvs - with a handle struct with methods, and maybe a more complete command syntax (or maybe just be transparent like iptables is).

thockin reviewed (commented):
Also, please squash the main commits together, but you can leave BUILD and lint as their own

prameshj commented:
Thanks for taking a look @thockin. I renamed the conntrack functions to the form ClearEntryFor* and added protocol as a parameter.
Squashed all changes into one commit, sorry could not separate out just the lint and BUILD changes.

thockin commented:
`git rebase -i` is your friend :)  No problem for now, but I strongly encourage you (and everyone) to learn it :)

thockin commented on pkg/kubelet/network/hostport/hostport_manager.go:
```diff
@@ -150,6 +165,21 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 		// clean up opened host port if encounter any error
 		return utilerrors.NewAggregate([]error{err, hm.closeHostports(hostportMappings)})
 	}
+	isIpv6 := conntrack.IsIPv6(podPortMapping.IP)
```
For future work - we should move the IsIPv6() stuff to a more generic place.

thockin commented on pkg/kubelet/network/hostport/hostport_manager.go:
```diff
@@ -150,6 +165,21 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
+	if hm.execer != nil && hm.conntrackFound {
+		glog.Infof("Starting to delete udp conntrack entries: %v, isIPv6 - %v", conntrackPortsToRemove, isIpv6)
+		for _, port := range conntrackPortsToRemove {
+			err = conntrack.ClearEntryForPort(hm.execer, port, isIpv6, v1.ProtocolUDP)
```
sorry to nit-pick.  "Entry" should be "Entries" I think?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -682,7 +683,7 @@ func (proxier *Proxier) deleteEndpointConnections(connectionMap []proxy.ServiceE
 		if svcInfo, ok := proxier.serviceMap[epSvcPair.ServicePortName]; ok && svcInfo.Protocol() == api.ProtocolUDP {
 			endpointIP := utilproxy.IPPart(epSvcPair.Endpoint)
-			err := utilproxy.ClearUDPConntrackForPeers(proxier.exec, svcInfo.ClusterIP(), endpointIP)
+			err := conntrack.ClearEntryForIPPair(proxier.exec, svcInfo.ClusterIP(), endpointIP, v1.ProtocolUDP)
```
IPPair is not quite descriptive - I would assume it is client/dest pair but it's really a NAT record.  ClearEntriesForNAT() ?

thockin commented on pkg/util/conntrack/conntrack.go:
```diff
@@ -14,42 +14,53 @@ See the License for the specific language governing permissions and
 	return IsIPv6(netIP)
 }
 
+func protoStr(proto v1.Protocol) string {
```
Why not just strings.ToLower() it and be done.  If the caller passes "ICMP", we should pass that thru to conntrack.

prameshj replied:
Good point, done. i left the protoStr function as is, in case we want to add any checks in future.  Incorporated the other comments as well.

prameshj commented:
I was trying to do a git rebase -i, but it was hard to isolate the BUILD and lint commits since there were some main files commits in between them as well. I will make sure I look into all options in the command. Thanks for reviewing the changes.

prameshj commented:
/test pull-kubernetes-e2e-kops-aws

dchen1107 commented:
/lgtm 

Are you going to send me a following pr to install the binary through cloud_init for GCE? That is not a good way to use cloud_init, but I guess I can live with it as a temporary workaround for now since we already did this for cni binaries.

prameshj commented:
@dchen1107 This is the PR to install the binary - https://github.com/kubernetes/kubernetes/pull/64640

subrnath commented:
I am facing similar issue with TCP https://github.com/kubernetes/kubernetes/issues/82727

AasthaRai95 commented:
I am still facing the issue. I have a service which collects the syslogs and runs on udp port. As soon as I delete the pod I am expecting that conntrack entries should get deleted but it doesnt delete on delete or even on restart. I am using kubernetes version - 
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.16", GitCommit:"e37e4ab4cc8dcda84f1344dda47a97bb1927d074", GitTreeState:"clean", BuildDate:"2021-10-27T16:25:59Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.9", GitCommit:"a1a87a0a2bcd605820920c6b0e618a8ab7d117d4", GitTreeState:"clean", BuildDate:"2023-04-12T12:08:36Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
What could be the problem here?

aojea commented:
hostPort depends on the runtime or plugin implementation, crio implements it in core and containerd use to rely in the cni portmap plugin

AasthaRai95 commented:
I didn't get you, could you please elaborate @aojea
